lib/RefBackend/RecognizeTorchFallbackPass.cpp

```diff
@@ -0,0 +1,101 @@
+//===- RecognizeTorchFallbackPass.cpp ---------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "npcomp/RefBackend/RefBackend.h"
+#include "PassDetail.h"
+
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "npcomp/Dialect/Basicpy/IR/BasicpyDialect.h"
+#include "npcomp/Dialect/Basicpy/IR/BasicpyOps.h"
+#include "npcomp/Dialect/Numpy/IR/NumpyDialect.h"
+#include "npcomp/Dialect/Numpy/IR/NumpyOps.h"
+#include "npcomp/Dialect/Refback/IR/RefbackOps.h"
+#include "npcomp/Dialect/Refback/IR/RefbackDialect.h"
+#include "npcomp/Dialect/Torch/IR/OpInterfaces.h"
+#include "npcomp/Dialect/Torch/IR/TorchOps.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "recognize-torch-fallback"
+
+using namespace mlir;
+using namespace mlir::NPCOMP;
+using namespace mlir::NPCOMP::refback;
+using namespace mlir::NPCOMP::Basicpy;
+using namespace mlir::NPCOMP::Torch;
+
+namespace {
+
+class EncapsulateKernelCallOpPattern : public OpRewritePattern<Torch::KernelCallOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(Torch::KernelCallOp kernelCall,
+                                PatternRewriter &rewriter) const override {
+    Operation *kcOp = kernelCall.getOperation();
+    // Change nothing if this KernelCallOp does not descend directly from a FuncOp.
```

Comment:
why?

Reply:
Because cloning the kernel-call leads to a recursive match that never stops; the kernel-call clone, which sits in its own new region, will also match. This is a base case that says only match kernel-calls at the top-level. Perhaps there is a more eloquent approach?

Reply:
In that case, check to see if we are already inside a fallback op, rather than if we are in a func op.